### PR TITLE
Reset per-SCC vertex marks in select_potentials (fixes #1201)

### LINKED ISSUE #1201: Soundness: select_potentials reports spurious infeasibility → meet returns bottom, vacuously passing assertions
## Summary

`select_potentials` (the per-SCC Bellman-Ford that assigns valid potentials during `meet`) can misreport a **satisfiable** constraint system as infeasible. When it does, `SplitDBM::meet` returns `nullopt` → bottom, and because `meet` sits directly in the fixpoint path, a spurious bottom at a loop head makes everything dominated by it "unreachable" — so every assertion there passes **vacuously**. This is a false-negative (accept-unsafe) soundness hole.

## Root cause

`src/crab/splitdbm/graph_ops.hpp:459-528`

Bellman-Ford runs per-SCC with a round budget of `|scc|`, but `vert_marks` are **never reset between SCCs** (only `compute_sccs` clears them once, up front). Relaxing a cross-SCC edge (`sd_pot < potentials[d]` at line 495) requeues a vertex `d` that belongs to an already-finished SCC, because its stale mark is still `BF_SCC` (line 497). That consumes the current (often size-1) SCC's iteration budget before the potentials converge, and the feasibility check at lines 513-524 then reads the still-relaxable leftover edge as a **negative cycle** and returns `false`.

Reached from `src/crab/splitdbm/split_dbm.cpp:388`:

```cpp
if (!select_potentials(*scratch_, result_g, result_pot)) {
    return std::nullopt; // Infeasible
}
```

## Why it affects soundness

`ZoneDomain::meet` → `SplitDBM::meet` is invoked in the extrapolator's fixpoint loop (`src/crab/extrapolator.cpp:27` `invariant = invariant & constant_limits_` and `:42` `invariant = invariant & new_pre`). A spurious bottom at a loop head vacuates all downstream assertion checks. (It also produces false "observation contradicts invariant" in reporting and silently bottoms test pre-invariants.)

## Traced counterexample

- left state `{x == y}` (edges x→y w0, y→x w0, potentials x=y=0)
- right state `{x − z ≤ −10}` (edge z→x w−10)
- `make_union_alignment` gives shared vars the **left** potentials, so x=y=z=0 initially.
- SCC order (Tarjan emits sinks-first): `{0}, {x,y}, {z}`.
- Processing `{z}` with budget 1: relax z→x sets `pot[x] = −10`, requeues `x` (stale mark `BF_SCC`); budget exhausted.
- Feasibility loop pops `x`, finds `x→y`: `−10 + 0 < pot[y]=0` → `return false`.

But `x == y ∧ x ≤ z − 10` is satisfiable, so `meet` wrongly returns bottom.

## Suggested fix

Restrict the infeasibility check to **intra-SCC** relaxations only, or reset `vert_marks` between SCCs / process SCCs in topological order so a cross-SCC relaxation cannot consume the current SCC's budget.

_Found in a static review. Static analysis only — no PoC executed, but the mechanism is confirmed by reading the code._

### PR DESCRIPTION
**Pull request by elazarg**
## Summary

Fixes #1201 — an accept-unsafe soundness hole where `select_potentials` reported a **satisfiable** constraint system as infeasible, making `SplitDBM::meet` return bottom and vacuously passing every assertion dominated by that program point.

## Root cause

`select_potentials` runs Bellman-Ford per SCC and uses `vert_marks` to identify vertices in the SCC currently being processed (`BF_SCC`), so that only intra-SCC relaxations requeue a vertex. But the marks were cleared only once up front and **never reset between SCCs**.

Tarjan emits SCCs sinks-first, so a source SCC is processed last. Relaxing one of its cross-SCC edges into an already-finished SCC lowered that vertex's potential and — because its stale `BF_SCC` mark still stood — requeued it. That consumed the current (often size-1) SCC's iteration budget before convergence, and the feasibility check then read the leftover relaxable cross-SCC edge as a **negative cycle**, returning `false`.

## Fix

A negative cycle is confined to a single SCC, so only intra-SCC relaxations are meaningful for feasibility. Clear each SCC's vertex marks once it is finished (matching the reference algorithm), so a cross-SCC relaxation neither requeues a finished vertex nor trips the feasibility check.

## Test

Regression test in `test_join.cpp` drives the issue's counterexample through `SplitDBM::meet` — left `{x == y}` meet right `{x - z <= -10}`. The meet is now non-bottom, and the closure correctly derives `y - z <= -10` (validating that the potentials remain usable end-to-end).

**Verified** the test spuriously bottoms (`meet` returns `nullopt`) without the fix and passes with it. Full suite: 636 passed, 1 skipped, 0 failures.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01QRunjNUunuZjp6KiBSZXzt

### CODE REVIEW

**Comment by elazarg** [comment]
@coderabbit review

**Comment by elazarg** [comment]
Thanks — both reviews are correct, and this was a real gap in the first commit, not just a stylistic one.

Clearing the per-SCC marks stopped the spurious bottom, but on its own it left `select_potentials` returning a **globally invalid** potential model: processing sink `{1,2}` before source `{3}` (Tarjan emits sinks-first), then relaxing `3→1` to `p[1]=-10` without revisiting `{1,2}`, leaves `p[2]=0` and violates the `1→2` edge. Those potentials are stored in the meet result and consumed by `chromatic_dijkstra`, which requires `p(d) <= p(s) + w`.

Fixed in 32baba5 by iterating the SCCs in **topological order** (walking Tarjan's reverse-topological output backwards), exactly as suggested. Now every cross-SCC edge points to a later, not-yet-processed SCC, so its relaxation tightens that SCC's seed before it runs and global validity is preserved; the mark reset is kept as defense-in-depth with an updated comment.

I also strengthened the regression test to assert the stored potentials are a valid model (`p(d) <= p(s) + w` for every edge). That assertion **failed** under the mark-reset-only fix (`0 <= -10` on the `x==y` edge) and passes with topological processing. Full suite green.

File: src/crab/splitdbm/graph_ops.hpp [hunk 1]
```diff
@@ -523,6 +523,17 @@ bool select_potentials(ScratchSpace& scratch, const ReadableGraph auto& g, std::
...
+
+        // Clear this SCC's vertex marks before moving on. Without this, the stale
+        // BF_SCC marks persist: relaxing a cross-SCC edge into an already-finished
+        // SCC would requeue one of its vertices (line "vert_marks.at(d) == BF_SCC"),
+        // consuming the current SCC's iteration budget and letting the feasibility
+        // check above read a leftover cross-SCC relaxation as a spurious negative
+        // cycle. A negative cycle is confined to a single SCC, so only intra-SCC
+        // relaxations are meaningful here. See issue #1201.
+        for (const VertId v : scc) {
+            scratch.vert_marks.at(v) = BF_NONE;
```
[the remark this answers is not in the document]
Comment (elazarg) [reply src/crab/splitdbm/graph_ops.hpp:542]: Should be fixed now
